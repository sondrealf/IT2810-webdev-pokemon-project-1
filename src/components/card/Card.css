.card__image-container {
  width: fit-content;
  margin: auto;
  margin-top: 25%;
}

.card__name-container {
  width: fit-content;
  margin: auto;
  padding-top: 5%;
  color: black;
  font-size: 1.5em;
  text-align: center;
}

.card__container {
  background-color: #fff;
  height: 300px;
  width: 200px;
  border-radius: 10px;
  margin: auto;
  font: inherit;
  display: block;
  border: none;
  padding: 0;
  color: black;
  transition: 200ms;
  box-shadow: 0px 0px 10px 2px rgba(0, 0, 0, 0.25);
  cursor: pointer;
}

.card__container:hover {
  transform: scale(1.05);
  box-shadow: 0px 0px 20px 4px rgba(0, 0, 0, 0.25);
}

.card__container:active {
  transform: scale(1.025);
  box-shadow: 0px 0px 15px 3px rgba(0, 0, 0, 0.25);
}

.card__header {
  display: flex;
  justify-content: space-evenly;
  gap: 0.5em;
  width: 100%;
}

.card__header-name {
  font-weight: 500;
}

.card__header-separator-line {
  height: 1px;
  background-color: #d2d2d2;
}

.card__image {
  height: 96px;
  width: 96px;
}
