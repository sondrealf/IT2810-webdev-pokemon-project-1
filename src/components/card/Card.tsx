import { useQuery } from "@tanstack/react-query";
import "./Card.css";
import "../../styles/App.css";
import PokemonType from "../../types/PokemonType";
import TypeCircle from "../typecircle/TypeCircle";
import FavouriteButton from "../favouriteButton/FavouriteButton";
import { useNavigate } from "react-router-dom";

/**
 * Custom hook that fetches a Pokemon by its ID.
 * @param id - The ID of the Pokemon to fetch.
 * @returns The result of the query, containing the Pokemon data.
 */
const useFetchPokemonQuery = (id: string) => {
  return useQuery<PokemonType>(["pokemon", id], async () =>
    (await fetch(id)).json()
  );
};

/**
 * Fetches data for a specific Pokemon based on the provided ID.
 * @param id - The ID of the Pokemon to fetch.
 * @returns - A React component that renders a card for a Pokemon.
 */
const Card = ({ id }: { id: string }) => {
  const { data, isError, isLoading } = useFetchPokemonQuery(id);
  const navigate = useNavigate();

  return (
    <button
      className="card__container"
      onClick={() => navigate(`/pokemon/${data?.id}`)}
    >
      <div className="card__header">
        {isLoading ? "" : isError ? "" : <p>{"#" + data?.id}</p>}
        <p className="card__header-name">
          {isError ? "Error" : isLoading ? "" : data.name.toUpperCase()}
        </p>
        {data?.types.length == 2 ? (
          <TypeCircle
            primaryType={data?.types[0].type.name}
            secondaryType={data?.types[1].type.name}
          />
        ) : (
          <TypeCircle primaryType={data?.types[0].type.name} />
        )}
      </div>

      <div className="card__header-separator-line" />

      <div className="card__image-container">
        {isError ? (
          "Error"
        ) : isLoading ? (
          <div className="card__image">{"Loading..."}</div>
        ) : data ? (
          <img
            className="card__image"
            src={data.sprites.front_default}
            alt={`Image of ${data.name}`}
          />
        ) : (
          "Something went wrong"
        )}
      </div>
      <FavouriteButton favID={data?.id ?? -1} />
    </button>
  );
};

export default Card;
