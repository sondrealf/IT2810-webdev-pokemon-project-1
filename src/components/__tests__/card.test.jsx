import React from "react";
import { render, waitFor, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import Card from "../card/Card.tsx";
import { it } from "vitest";

const mockUseNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockUseNavigate,
  };
});

describe("Card component", () => {
  const mockData = {
    id: 1,
    name: "bulbasaur",
    types: [{ type: { name: "grass" } }, { type: { name: "poison" } }],
    sprites: { front_default: "bulbasaur.png" },
  };

  const mockFetch = vi.fn(() => Promise.resolve({ json: () => mockData }));

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  const renderComponent = (props) =>
    render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <Card {...props} />
        </MemoryRouter>
      </QueryClientProvider>
    );

  beforeEach(() => {
    global.fetch = mockFetch;
  });

  afterEach(() => {
    vi.clearAllMocks();
    global.fetch = undefined;
  });

  it("matches snapshot", async () => {
    const { asFragment } = renderComponent({ id: "1" }); //TODO
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(1));
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders pokemon name and number", async () => {
    renderComponent({ id: "1" }); //TODO
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(1));
    expect(screen.getByText("#1")).toBeInTheDocument();
    expect(screen.getByText("BULBASAUR")).toBeInTheDocument();
  });

  it("renders pokemon image", async () => {
    renderComponent({ id: "1" });
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(1));
    expect(screen.getByRole("img")).toHaveAttribute("src", "bulbasaur.png");
  });

  it("navigates to pokemon details page when clicked", async () => {
    renderComponent({ id: "1" });
    screen.getByRole("img").click();
    expect(mockUseNavigate).toHaveBeenCalledWith("/pokemon/1");
    expect(mockUseNavigate).toHaveBeenCalledTimes(1);
  });
});
