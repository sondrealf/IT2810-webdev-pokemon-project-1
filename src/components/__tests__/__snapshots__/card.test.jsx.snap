// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="card__container"
  >
    <div
      class="card__header"
    >
      <p
        class="card__header-name"
      />
      <div
        class="TypeCircle__container"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 16 15"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="TypeCircle__typeColor-undefined"
            d="M7.50488 15C5.51576 15 3.6081 14.2098 2.20158 12.8033C0.795059 11.3968 0.00488281 9.48912 0.00488281 7.5C0.00488281 5.51088 0.795059 3.60322 2.20158 2.1967C3.6081 0.790177 5.51576 3.00349e-07 7.50488 0L7.50488 7.5V15Z"
          />
          <path
            class="TypeCircle__typeColor-undefined"
            d="M7.50488 0C9.49401 0 11.4017 0.790176 12.8082 2.1967C14.2147 3.60322 15.0049 5.51088 15.0049 7.5C15.0049 9.48912 14.2147 11.3968 12.8082 12.8033C11.4017 14.2098 9.49401 15 7.50488 15L7.50488 7.5L7.50488 0Z"
          />
        </svg>
      </div>
    </div>
    <div
      class="card__header-separator-line"
    />
    <div
      class="card__image-container"
    >
      <div
        class="card__image"
      >
        Loading...
      </div>
    </div>
    <div
      class="FavouriteButton__container"
    >
      <svg
        height="25"
        role="svg"
        viewBox="0 0 22 23"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=" FavouriteButton__starpath"
          d="M8.77229 2.22466C9.47249 -0.174957 12.8719 -0.174952 13.5721 2.22467L14.2763 4.63798C14.5875 5.7045 15.5652 6.43769 16.6762 6.43769H19.481C21.8374 6.43769 22.8853 9.39923 21.0534 10.8813L18.3381 13.0781C17.5545 13.712 17.2282 14.7544 17.5106 15.7219L18.4363 18.8947C19.1225 21.2461 16.3683 23.0792 14.464 21.5385L12.7446 20.1475C11.8276 19.4056 10.5168 19.4056 9.59981 20.1475L7.8804 21.5385C5.97608 23.0792 3.22194 21.2461 3.90807 18.8947L4.83386 15.7219C5.11618 14.7544 4.7899 13.712 4.00635 13.0781L1.291 10.8813C-0.540912 9.39924 0.507053 6.43769 2.8634 6.43769H5.66818C6.77918 6.43769 7.7569 5.7045 8.0681 4.63798L8.77229 2.22466Z"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;
